### PR TITLE
Products: Silence parsing errors to unblock Products tab and log errors instead

### LINKED ISSUE #7293: Products: Make parsing fail silently for ProductImage's `alt` and Product's `variations`
# The problem
As reported in 5357954-zd-woothemes - a merchant with an incorrect type for the `alt` field of product images cannot load the product list since the decoding fails:

```
⛔️ Error synchronizing products: typeMismatch(Swift.String, Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "data", intValue: nil), _JSONKey(stringValue: "Index 2", intValue: 2), CodingKeys(stringValue: "images", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "alt", intValue: nil)], debugDescription: "Expected to decode String but found a number instead.", underlyingError: nil))
```
Also, the `variations` field of `Product` returns a list of variation objects instead of expected list of variation IDs:

```
Error synchronizing products: typeMismatch(Swift.Int64, Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "data", intValue: nil), _JSONKey(stringValue: "Index 9", intValue: 9), CodingKeys(stringValue: "variations", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0)], debugDescription: "Expected to decode Int64 but found a dictionary instead.", underlyingError: nil))
```

Potential plugins that cause this:
- Variation Swatches for WooCommerce
- Additional Variation Images Gallery for WooCommerce

# Expectation
We want the Products tab to be accessible, so:
- Since `alt` is an optional field, make the decoding fail silently if it's impossible to decode the field as a `String`.
- If `variations` fails to be parsed, return an empty list instead.

Internal discussion: p1657815453917059-slack-C6H8C3G23

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #7293 

### Description

As reported in 5357954-zd-woothemes - a merchant with an incorrect type for the `alt` field of product images and  `variations` of products cannot load the product list since the decoding fails.

This PR attempts to unblock the Product tab by letting the parsing error go silent while logging the errors for tracking the issue if needed.

### Testing instructions

I cannot detect which plugin causes the malformed response, so I'm adding extra unit tests for the model mappers.
Please feel free to test with a normal store that has product variations and images to make sure that every thing still works properly.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Model/Product/Product.swift [hunk 1]
```diff
@@ -380,7 +380,14 @@ public struct Product: Codable, GeneratedCopiable, Equatable, GeneratedFakeable
 
         let attributes = try container.decode([ProductAttribute].self, forKey: .attributes)
         let defaultAttributes = try container.decode([ProductDefaultAttribute].self, forKey: .defaultAttributes)
-        let variations = try container.decode([Int64].self, forKey: .variations)
+        let variations: [Int64] = {
+            do {
+                return try container.decode([Int64].self, forKey: .variations)
+            } catch {
+                DDLogError("⛔️ Error parsing `variations` for `Product`: \(error)")
```
Comment (iamgabrielma) [inline Networking/Networking/Model/Product/Product.swift]: Would it be useful to pass the productID here so is easier to find out the product that is logging the error? 

```
DDLogError("⛔️ Error parsing `variations` for `Product` \(productID): \(error)")
```

I'm not entirely sure of what data the error will contain as I was unable to replicate this.

File: Networking/Networking/Model/Product/ProductImage.swift [hunk 2]
```diff
@@ -37,7 +37,14 @@ public struct ProductImage: Codable, Equatable, GeneratedCopiable, GeneratedFake
         let dateModified = try container.decodeIfPresent(Date.self, forKey: .dateModified)
         let src = try container.decode(String.self, forKey: .src)
         let name = try container.decodeIfPresent(String.self, forKey: .name)
-        let alt = try container.decodeIfPresent(String.self, forKey: .alt)
+        let alt: String? = {
+            do {
+                return try container.decodeIfPresent(String.self, forKey: .alt)
+            } catch {
+                DDLogError("⛔️ Error parsing `alt` for `ProductImage`: \(error)")
```
Comment (iamgabrielma) [inline Networking/Networking/Model/Product/ProductImage.swift]: Same as before, would be useful to pass the `imageID`? 
```
DDLogError("⛔️ Error parsing `alt` for `ProductImage` \(imageID): \(error)")
```

**Review by iamgabrielma — APPROVED** [review]
LGTM! I wasn't able to replicate this either, despite using different variations plugins, but by attempting to break the malformed response further your changes always worked nicely and tests keep passing 💯 

:shipit:

File: Networking/Networking/Model/Product/Product.swift (hunk 1 above)
Comment (itsmeichigo) [reply Networking/Networking/Model/Product/Product.swift]: The error will look like this:

```
typeMismatch(Swift.String, Swift.DecodingError.Context(codingPath: [CodingKeys(stringValue: "data", intValue: nil), _JSONKey(stringValue: "Index 2", intValue: 2), CodingKeys(stringValue: "images", intValue: nil), _JSONKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "alt", intValue: nil)], debugDescription: "Expected to decode String but found a number instead.", underlyingError: nil))
```

Also, good point about tracking the product ID! That'll be helpful for troubleshooting. Updated in 2bed57dd321e45b2c8b08fc6485cd21c87497071.

File: Networking/Networking/Model/Product/ProductImage.swift (hunk 2 above)
Comment (itsmeichigo) [reply Networking/Networking/Model/Product/ProductImage.swift]: Also updated in 2bed57dd321e45b2c8b08fc6485cd21c87497071 👍

**Comment by thelogicwizards** [comment]
User in [5504447-ZD](https://woothemes.zendesk.com/agent/tickets/5504447) reported that their Orders aren't loading since the update although Products are loading as expected.

**Comment by itsmeichigo** [comment]
@thelogicwizards I cannot access the ticket at the moment, but please file a separate issue based on the logs for the order sync requests if possible, thank you.

**Pull request merged**